File: 栈/155.最小栈.py

# (val, now_min_val) 保存当前值和当前最小值
class MinStack(object):

    def __init__(self):

        self.stack = []

    def push(self, x):
        if not self.stack:
            self.stack.append((x, x))
        else:
            self.stack.append((x, min(x, self.stack[-1][1])))

    def pop(self):

        self.stack.pop()

    def top(self):

        return self.stack[-1][0]

    def getMin(self):

        return self.stack[-1][1]

File: 栈/155.最小栈.js
var MinStack = function () {
  this.stack = [];
};

/**
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
  if (!this.stack.length) this.stack.push([val, val]);
  else
    this.stack.push([val, Math.min(val, this.stack[this.stack.length - 1][1])]);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  return this.stack.pop()[0];
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  return this.stack[this.stack.length - 1][0];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  return this.stack[this.stack.length - 1][1];
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
